File: ast_canopy/cpp/src/enum.cpp
// clang-format off
// SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0
// clang-format on

#include <ast_canopy/ast_canopy.hpp>

namespace ast_canopy {

Enum::Enum(const clang::EnumDecl *ED)
    : name(ED->getNameAsString()),
      underlying_type(ED->getIntegerType(), ED->getASTContext()),
      qual_name(ED->getQualifiedNameAsString()) {
  if (qual_name.empty()) {
    qual_name = name;
  }
  for (const auto *enumerator : ED->enumerators()) {
    enumerators.push_back(enumerator->getNameAsString());

    auto const val = enumerator->getInitVal();
    llvm::SmallVector<char> buf;
    val.toString(buf);
    std::string s(buf.begin(), buf.end());

    enumerator_values.push_back(s);
  }
}

} // namespace ast_canopy

File: numbast/src/numbast/enum.py
# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0

from enum import IntEnum

from ast_canopy import pylibastcanopy as pylibastcanopy
from numbast.types import register_enum_type


def bind_cxx_enum(e: pylibastcanopy.Enum):
    if len(e.enumerator_values) == 0:
        return None

    vals = [int(v) for v in e.enumerator_values]
    # Numba takes python enum object as-is. Thus we only need to dynamically create
    # the enum object and return it.
    pyenum = IntEnum(e.name, dict(zip(e.enumerators, vals)))

    example_value = vals[0]

    # Add the enum to the CTYPE_MAPS
    register_enum_type(e.name, pyenum(example_value))

    return pyenum


def bind_cxx_enums(enums: list[pylibastcanopy.Enum]) -> list[IntEnum]:
    """
    Create bindings for a list of C++ enums.

    Parameters
    ----------
    enums : list[pylibastcanopy.Enum]
        A list of enum declarations in CXX.

    Returns
    -------
    pyenums : list[IntEnum]
    """

    pyenums = []

    for e in enums:
        pyenum = bind_cxx_enum(e)
        if pyenum is not None:
            pyenums.append(pyenum)

    return pyenums
